src/commonMain/kotlin/com/github/quillraven/fleks/collection/bitArray.kt

```diff
@@ -36,30 +36,31 @@ class BitArray(
         get() = !isNotEmpty
 
     operator fun get(idx: Int): Boolean {
-        val word = idx / 64
-        return if (word >= bits.size) {
-            false
-        } else {
-            (bits[word] and (1L shl (idx % 64))) != 0L
-        }
+        val word = idx ushr 6
+        return word < bits.size && (bits[word] and (1L shl idx)) != 0L
     }
 
     fun set(idx: Int) {
-        val word = idx / 64
+        val word = idx ushr 6
         if (word >= bits.size) {
-            bits = bits.copyOf(word + 1)
+            // magic.
```

Comment:
I have some questions here:
- why is this approach better (and faster?) then a simple `word + 1`?
- Your approach always created a bits array with a size that is a multiple of 2. I could imagine this has a positive effect as well since normally multiple of 2 things are faster?


Otherwise, I suggest to remove `magic.` and the `With the rest....` comment since your other comment explains it very well and at least I don't care about the comparison of a previous approach. We anyway have a git history, if someone is interested ;)

Reply:
Part 1: So the reason `word + 1` might be slower is that it only resizes a little bit at a time -- so if you start at size 1 and gradually set() larger values, it has to allocate a new bits Array every time you set a larger value. This can be an issue for speed initially, though not a big one, but it's a bigger problem for garbage collection, since this could create quite a few Array items in one frame, all but one of which need GC-ing. By doubling the array size, the total number of resizes possible for an Array maxes out at 31 or less (the maximum size of an Array on the JVM is just shy of 2 to the 31, about 2.1 billion); when it is already setting inside the maximum size, no resize is needed. It does use more memory as a maximum for data that isn't being modified, but it discards far fewer Arrays along the way. If you know your data only grows to small maximum sizes, then adding one to the size isn't that bad, and it ensures optimal memory usage when not setting any bits. This is a pretty classic tradeoff between best-case memory usage and best-case (and average-case) speed. To me, the stuttering associated with garbage collection makes me really try to avoid it in game code.

Part 2: I know it isn't clear code; that's why I tried to comment it. I wasn't comparing the previous approach at all; I was describing the resize logic in terms of before resize and after resize.

Addendum: It may be worth benchmarking if this equivalent code is faster:

`bits = bits.copyOf(max(word + word, 1))`

The performance of max() varies across JVM versions, but the bitwise code used now should be stable. The bitwise code is also branchless, which doesn't have as much effect as one would think due to speculative execution and such. I'm not sure if there's some compiler code that makes max() branchless or effectively so.